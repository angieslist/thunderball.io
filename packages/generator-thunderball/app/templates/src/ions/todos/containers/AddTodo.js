import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addTodo } from '../actions';

function AddTodo({ dispatch }) {
  const [value, setValue] = React.useState('');
  const handleChange = event => setValue(event.target.value);
  const handleSubmit = (event) => {
    event.preventDefault();
    if (!value.trim()) {
      return;
    }
    dispatch(addTodo(value));
    setValue('');
  };
  return (
    <div>
      <form onSubmit={handleSubmit}>
        <input value={value} onChange={handleChange} />
        <button type="submit">
          Add Todo
        </button>
      </form>
    </div>
  );
}

AddTodo.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

export default connect()(AddTodo);
