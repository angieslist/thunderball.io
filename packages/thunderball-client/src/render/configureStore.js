import appReducers from '../reducers/index';
import { createLogger } from 'redux-logger';
import { applyMiddleware, createStore, compose, combineReducers } from 'redux';
import handleResetOnAction from './handleResetOnAction';

const isReactNative = typeof navigator === 'object' && navigator.product === 'ReactNative';
const defaultStoreOptions = {
  middleware: [],
  enhancers: [],
  reducers: appReducers,
  initialState: {},
  enableLogger: process.env.NODE_ENV !== 'production' && process.env.IS_BROWSER || isReactNative,
  resetOnAction: [],
};

export default function configureStore(options = {}, injectors = [], pageProps = {}) {
  // Injection point: beforeConfigureStore
  // Using the beforeConfigureStore, determine what the full storeOptions should be
  // by reducing each result of beforeConfigureStore with the previous
  const storeOptions = injectors
    .filter(injector => injector.beforeConfigureStore)
    .reduce((prevOptions, injector) => ({ ...prevOptions, ...injector.beforeConfigureStore(prevOptions, pageProps) }),
      // Create the base storeOptions from the defaults and those defined by the platform
      { ...defaultStoreOptions, ...options });

  // Logger must be the last middleware in chain.
  if (storeOptions.enableLogger) {
    storeOptions.middleware.push(createLogger({
      collapsed: true,
      // Convert objects to pure javascript
      stateTransformer: state => JSON.parse(JSON.stringify(state)),
    }));
  }

  // Add middleware as the first enhancer
  storeOptions.enhancers.unshift(applyMiddleware(...storeOptions.middleware));

  // Always push reduxDevTools to the end of the enhancer stack
  // This is only active in development mode and in clients with the devtools installed
  // Redux tools extension https://github.com/zalmoxisus/redux-devtools-extension
  if (process.env.NODE_ENV !== 'production' &&
    typeof window === 'object' && typeof window.devToolsExtension === 'function') {
    storeOptions.enhancers.push(window.devToolsExtension());
  }

  // Combine reducers and create store
  const combinedReducer = combineReducers(storeOptions.reducers);
  const store = createStore(
    handleResetOnAction(combinedReducer, storeOptions.initialState, storeOptions.resetOnAction),
    storeOptions.initialState,
    compose(...storeOptions.enhancers),
  );

  // Injection point: afterConfigureStore
  injectors
    .filter(injector => injector.afterConfigureStore)
    .forEach(injector => injector.afterConfigureStore(store, pageProps, storeOptions));

  return store;
}
