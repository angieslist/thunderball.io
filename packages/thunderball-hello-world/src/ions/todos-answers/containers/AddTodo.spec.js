import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';

import AddTodo from './AddTodo';

const setup = () => {
  const store = configureStore()({ todos: [] });
  const wrapper = shallow(<AddTodo store={store} />);

  return {
    store,
    wrapper,
  };
};

describe('AddTodo', () => {
  test('renders without crashing', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  test('onSubmit calls preventDefault to stop form submit', () => {
    const preventDefault = jest.fn();
    const { wrapper } = setup();
    const form = wrapper.shallow().shallow().find('form');
    form.simulate('submit', { preventDefault });
    expect(preventDefault).toHaveBeenCalled();
  });

  test('onSubmit returns early if input.value is not set', () => {
    const { store } = setup();
    expect(store.getActions()).toEqual([]);
  });

  test('onSubmit dispatches action if input.value is set', () => {
    const { store, wrapper } = setup();
    const preventDefault = jest.fn();
    const deeperWrapper = wrapper.shallow();

    deeperWrapper
      .find('input')
      .simulate('change', { target: { value: 'Test todo' } });

    deeperWrapper.find('form').simulate('submit', { preventDefault });

    expect(preventDefault).toHaveBeenCalled();
    expect(store.getActions()).toMatchObject([{
      id: expect.any(Number),
      text: 'Test todo',
      type: 'ADD_TODO',
    }]);
  });
});
