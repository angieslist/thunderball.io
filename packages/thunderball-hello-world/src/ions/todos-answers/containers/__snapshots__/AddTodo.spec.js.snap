// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo renders without crashing 1`] = `
<AddTodo
  dispatch={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
